src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -545,5 +543,47 @@ internal void TakeSnapshot(bool outputDelta)
         {
             throw new NotSupportedException($"{methodName} is not supported for this metric type.");
         }
+
+#pragma warning disable SA1204 // Static elements should appear before instance elements
+        private static int FindHistogramBucketIndexLinear(double[] bounds, double number)
+#pragma warning restore SA1204 // Static elements should appear before instance elements
+        {
+            int i;
+
+            for (i = 0; i < bounds.Length; i++)
+            {
+                if (number <= bounds[i])
+                {
+                    break;
+                }
+            }
+
+            return i;
+        }
+
+        private static int FindHistogramBucketIndexBinary(double[] bounds, double number)
+        {
+            var left = 0;
+            var right = bounds.Length - 1;
+
+            while (left <= right)
+            {
+                var mid = (int)Math.Floor((double)(left + right) / 2);
```

Comment:
Is there a way to avoid conversion to double? (I guess it's expensive)

Reply:
What's the reason behind implementing the BinarySearch algorithm and not using `Array.BinarySearch`. If I remember correctly you had a created a PR long ago where you used `Array.BinarySearch`?

Reply:
IIRC `Array.BinarySearch` was slower since it has to work with all different types and be safer. I can check its perf again to be sure

Reply:
The results seem comparable. Wouldn't it be better to use `Array.BinarySearch` and avoid additional maintenance?

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3252#discussion_r872582866